### ISSUE TITLE
Add NavigationListeners that can be registered manually

### ISSUE THREAD

**Issue by pleku** [vaadin/flow#2754]
After #2737 we have `BeforeNavigationObserver`, `BeforeEnterObserver` and `BeforeLeaveObserver` which are automatically registered. We should add the corresponding `Listener` interfaces, which the observers should extend, and it needs to be possible to manually register (and unregister) those listeners for a specific UI.

**Comment by caalador** [comment]
[NOTE]
`BeforeNavigationObserver`has been removed.

**Comment by heruan** [comment]
@berndhopp Is the PR still work in progress?

**Comment by Legioth** [comment]
I'm not entirely sure which case makes more sense:
* Adding listeners for a specific UI instance.
* Adding listeners to e.g. `VaadinService` that are triggered for all UI instances in the application.
* Maybe even both?

**Comment by caalador** [comment]
Probably both. I sort of understand this case as being able to decide on needed listeners on UI init, where as with the `VaadinServiceInitListener`type of registration every one would always get all the listeners.

**Comment by Legioth** [comment]
What use cases are we aware of for per-UI listeners?

For a global listener, one use case it at least authorization that might want to reject opening a view depending on the permissions of the currently authenticated user.

Does @berndhopp who made the original patch have any input on the expected use cases?

**Comment by caalador** [comment]
He had in the original PR `use-case for this PR` [here](https://github.com/berndhopp/vaadin-security/blob/v10/src/main/java/org/ilay/Ilay.java)

But that looks like it should be for a global navigationEvent listener.

**Comment by Legioth** [comment]
Written by @berndhopp in the PR review:
> since I was the author of the original PR that this one evolved from, I feel authorized to answer that question.
The problem to be solved is to bring back something similar to a ViewChangeListener in V8. ViewChangeListeners by nature lived in a 'UI-Scope' since the navigator was bound to a specific UI. If we now change that scope to 'singleton', a single listener for 'all' navigations would need to be stateless, since it deals with often times sensitive session-scoped information like login, credentials etc.. And if people port over their ViewChangeListeners from V8, it would make things a lot harder for them if they would be forced to use a singleton now.

**Comment by Legioth** [comment]
I think the key here is that the listener would deal with session-scoped information, but I don't see why that would mean that the listener itself would have to directly contain any such state.

Sure, it would be convenient to be able to reuse exactly the same structure when migrating, but I'm not sure that makes very much sense in this case since all other aspects of navigation/routing has also changed to be either "singleton" or tied to individual attached component instances.

**Comment by berndhopp** [comment]
but what are the benefits of having a singleton listener, besides that we wouldn't break the 'singleton-or-attached-component'-pattern this way?

**Comment by Legioth** [comment]
Two benefits: Simpler registration (you don't need to put an UI attach listener in between) and less memory usage (since there's only one instance).

**Comment by berndhopp** [comment]
on the other hand, I would argue it's more complicated and error-prone when these listeners are not in the UI-scope. For example, if one would annotate such a listener with @UIScope in Spring or Guice, the results are pretty much undefined if the listener is by nature in Singleton-Scope. 
Memory-usage will be slightly higher, but to a negligible scale in my assumption, since there will be no heavyweight objects introduced.

**Comment by Legioth** [comment]
Yes, things would work in weird ways if using an inappropriate scope in this case, just as in any other case. Why would one annotate a globally registered listener type with `@UIScope` in the first place, other than if trying to directly port something from previous framework versions?

**Comment by berndhopp** [comment]
well, for exactly that reason, backwards-portability is one of my main concerns here. another case would be when DI-frameworks get support for these listeners and pick them up automatically ( think of package-scan ). Then when somebody puts the listener in anything but singleton-scope, that would be a undefined state with unpredictable behavior.

**Comment by berndhopp** [comment]
basically, to me the 'singleton-or-attached-component'-pattern is limiting here, because not everything is either a singleton or a component.

**Comment by berndhopp** [comment]
but I give you that is not a 'hard blocker', if you want to have a singleton-listener ( or a singleton list of listeners ), I'm fine with it as long as it is not being postponed forever now, because we need that feature in the you-know-which-project :-)

**Comment by Legioth** [comment]
I don't see either option as more or less limiting since one option can easily be implemented on top of the other with a reasonable amount of boilerplate: You can have one globally registered listener instance that looks up a UI-specific instance and delegates to it, or you can manually register the same global instance with each separate UI instance.

The DI integration case with global registration and bad scope annotations would assumingly cause a quite clear exception from the DI framework when there wouldn't be any UI scope active when the registration logic would try to create an instance. Alternatively, a more advanced DI integration could explicitly check the scope of the type and treat it differently depending on the case.

Identified benefits for each approach:
* Global registration
  * More straightforward registration when no UI attach listener is needed
  * Less memory usage since there's only one instance in the whole application
* Per-UI registration
  * More familiar when migrating from previous versions
  * Enables capturing UI-specific in the listener instance itself

It feels like there is no ultimate solution, which means that we might consider having support for both (in the same way that V8 `RequestHandler` can be registered both on the `VaadinService` or `VaadinSesison` level). This in turn would mean that the current patch could go in as-is, and then the other approach could be implemented separately at any point.

**Comment by berndhopp** [comment]
agreed
